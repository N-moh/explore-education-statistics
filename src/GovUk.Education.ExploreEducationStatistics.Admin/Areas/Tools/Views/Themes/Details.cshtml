@model GovUk.Education.ExploreEducationStatistics.Content.Model.Theme

@{
    ViewData["Title"] = Model.Title;
}

<span class="govuk-caption-l">Theme</span>
<h1 class="govuk-heading-l">@Model.Title</h1>


<div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
        <h3 class="govuk-heading-m">Details</h3>

        <table class="govuk-table">
            <tbody>
            <tr class="govuk-table__row">
                <td class="govuk-table__header">
                    @Html.DisplayNameFor(model => model.Slug)
                </td>
                <td class="govuk-table__cell">
                    /@Html.DisplayFor(model => model.Slug)/
                </td>
            </tr>
            <tr class="govuk-table__row">
                <td class="govuk-table__header">
                    @Html.DisplayNameFor(model => model.Summary)
                </td>
                <td class="govuk-table__cell">
                    @Html.DisplayFor(model => model.Summary)
                </td>
            </tr>
            </tbody>
        </table>

    </div>

    <div class="govuk-grid-column-one-third">
        <h3 class="govuk-heading-m">Topics</h3>
        <ul class="govuk-list">
            @foreach (var item in Model.Topics)
            {
                <li>
                    <a asp-route-id="@item.Id" asp-action="Details" asp-controller="Topics" class="govuk-link">@Html.DisplayFor(modelItem => item.Title)</a>
                </li>
            }
        </ul>

    </div>
</div>

<div>
    <a asp-action="Edit" asp-route-id="@Model.Id" class="govuk-button">Edit</a>
    <p>
        <a asp-action="Index" class="govuk-back-link">Back to List</a>
    </p>
</div>