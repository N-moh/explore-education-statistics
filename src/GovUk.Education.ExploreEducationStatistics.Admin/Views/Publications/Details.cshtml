@model GovUk.Education.ExploreEducationStatistics.Content.Model.Publication

@{
    ViewData["Title"] = Model.Title;
}

<span class="govuk-caption-l">Publication</span>
<h1 class="govuk-heading-l">@Model.Title</h1>

<div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
        <h3 class="govuk-heading-m">Details</h3>

        <table class="govuk-table">
            <tbody>
            <tr class="govuk-table__row">
                <td class="govuk-table__header">
                    @Html.DisplayNameFor(model => model.Topic)
                </td>
                <td class="govuk-table__cell">
                    <a asp-controller="Topics" asp-action="Details" asp-route-id="@Model.TopicId" class="govuk-link">@Html.DisplayFor(model => model.Topic.Title)</a>
                </td>
            </tr>
            <tr class="govuk-table__row">
                <td class="govuk-table__header">
                    @Html.DisplayNameFor(model => model.Slug)
                </td>
                <td class="govuk-table__cell">
                    /@Html.DisplayFor(model => model.Slug)/
                </td>
            </tr>
            <tr class="govuk-table__row">
                <td class="govuk-table__header">
                    @Html.DisplayNameFor(model => model.Summary)
                </td>
                <td class="govuk-table__cell">
                    @Html.DisplayFor(model => model.Summary)
                </td>
            </tr>

            <tr class="govuk-table__row">
                <td class="govuk-table__header">
                    @Html.DisplayNameFor(model => model.NextUpdate)
                </td>
                <td class="govuk-table__cell">
                    @Html.DisplayFor(model => model.NextUpdate)
                </td>
            </tr>
            <tr class="govuk-table__row">
                <td class="govuk-table__header">
                    @Html.DisplayNameFor(model => model.LegacyPublicationUrl)
                </td>
                <td class="govuk-table__cell">
                    <a href="@Model.LegacyPublicationUrl" class="govuk-link" target="_blank">@Html.DisplayFor(model => model.LegacyPublicationUrl)</a>
                </td>
            </tr>
            </tbody>
        </table>
    </div>

    <div class="govuk-grid-column-one-third">
        <h3 class="govuk-heading-m">Releases</h3>
        <ul class="govuk-list">
            @foreach (var item in Model.Releases)
            {
                <li>
                    <a asp-route-id="@item.Id" asp-action="Details" asp-controller="Publications" class="govuk-link">@Html.DisplayFor(modelItem => item.Title)</a>
                </li>
            }
        </ul>
    </div>

</div>
<div>
    <a asp-action="Edit" asp-route-id="@Model.Id" class="govuk-button">Edit</a>
    <p>
        <a asp-action="Index" class="govuk-back-link">Back to List</a>
    </p>
</div>