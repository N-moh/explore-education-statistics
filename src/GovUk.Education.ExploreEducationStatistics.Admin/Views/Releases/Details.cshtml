@model GovUk.Education.ExploreEducationStatistics.Content.Model.Release

@{
    ViewData["Title"] = "Details";
}

<h1>Details</h1>

<div>
    <h4>Release</h4>
    <hr />
    <dl class="row">
        <dt class = "col-sm-2">
            @Html.DisplayNameFor(model => model.Title)
        </dt>
        <dd class = "col-sm-10">
            @Html.DisplayFor(model => model.Title)
        </dd>
        <dt class = "col-sm-2">
            @Html.DisplayNameFor(model => model.ReleaseName)
        </dt>
        <dd class = "col-sm-10">
            @Html.DisplayFor(model => model.ReleaseName)
        </dd>
        <dt class = "col-sm-2">
            @Html.DisplayNameFor(model => model.Published)
        </dt>
        <dd class = "col-sm-10">
            @Html.DisplayFor(model => model.Published)
        </dd>
        <dt class = "col-sm-2">
            @Html.DisplayNameFor(model => model.Slug)
        </dt>
        <dd class = "col-sm-10">
            @Html.DisplayFor(model => model.Slug)
        </dd>
        <dt class = "col-sm-2">
            @Html.DisplayNameFor(model => model.Summary)
        </dt>
        <dd class = "col-sm-10">
            @Html.DisplayFor(model => model.Summary)
        </dd>
        <dt class = "col-sm-2">
            @Html.DisplayNameFor(model => model.Publication)
        </dt>
        <dd class = "col-sm-10">
            @Html.DisplayFor(model => model.Publication.Title)
        </dd>
        <dt class = "col-sm-2">
            @Html.DisplayNameFor(model => model.Content)
        </dt>
        <dd class = "col-sm-10">
            @Html.DisplayFor(model => model.Content)
        </dd>
        <dt class = "col-sm-2">
            @Html.DisplayNameFor(model => model.KeyStatistics)
        </dt>
        <dd class = "col-sm-10">
            @Html.DisplayFor(model => model.KeyStatistics)
        </dd>
    </dl>
</div>
<div>
    <a asp-action="Edit" asp-route-id="@Model.Id">Edit</a> |
    <a asp-action="Index">Back to List</a>
</div>
