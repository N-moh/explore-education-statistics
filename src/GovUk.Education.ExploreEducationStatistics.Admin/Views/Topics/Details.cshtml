@model GovUk.Education.ExploreEducationStatistics.Content.Model.Topic

@{
    ViewData["Title"] = Model.Title;
}
<span class="govuk-caption-l">Topic</span>
<h1 class="govuk-heading-l">@Model.Title</h1>

<div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
        <h3 class="govuk-heading-m">Details</h3>

        <table class="govuk-table">
            <tbody>
            <tr class="govuk-table__row">
                <td class="govuk-table__header">
                    @Html.DisplayNameFor(model => model.Theme)
                </td>
                <td class="govuk-table__cell">
                    @Html.DisplayFor(model => model.Theme.Title)
                </td>
            </tr>
            <tr class="govuk-table__row">
                <td class="govuk-table__header">
                    @Html.DisplayNameFor(model => model.Slug)
                </td>
                <td class="govuk-table__cell">
                    /@Html.DisplayFor(model => model.Slug)/
                </td>
            </tr>
            <tr class="govuk-table__row">
                <td class="govuk-table__header">
                    @Html.DisplayNameFor(model => model.Summary)
                </td>
                <td class="govuk-table__cell">
                    @Html.DisplayFor(model => model.Summary)
                </td>
            </tr>
            </tbody>
        </table>
    </div>
    <div class="govuk-grid-column-one-third">
        <h3 class="govuk-heading-m">Publications</h3>
        <ul class="govuk-list">
            @foreach (var item in Model.Publications)
            {
                <li>
                    <a asp-route-id="@item.Id" asp-action="Details" asp-controller="Publications" class="govuk-link">@Html.DisplayFor(modelItem => item.Title)</a>
                </li>
            }
        </ul>
    </div>

</div>

<div>
    <a asp-action="Edit" asp-route-id="@Model.Id" class="govuk-button">Edit</a>
    <p>
        <a asp-action="Index" class="govuk-back-link">Back to List</a>
    </p>
</div>