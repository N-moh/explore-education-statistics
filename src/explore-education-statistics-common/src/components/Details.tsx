import useMounted from '@common/hooks/useMounted';
import useToggle from '@common/hooks/useToggle';
import findAllParents from '@common/lib/dom/findAllParents';
import classNames from 'classnames';
import React, { MouseEvent, ReactNode, useEffect, useRef } from 'react';

let hasNativeDetails: boolean;
let idCounter = 0;

export interface DetailsProps {
  className?: string;
  tag?: string;
  children: ReactNode;
  id?: string;
  onToggle?: (isOpen: boolean, event: MouseEvent<HTMLElement>) => void;
  analytics?: (eventGA: { category: string; action: string }) => void;
  open?: boolean;
  summary: string | ReactNode;
}

const Details = ({
  className,
  children,
  id = `details-content-${(idCounter += 1)}`,
  open = false,
  onToggle,
  analytics,
  summary,
  tag,
}: DetailsProps) => {
  const ref = useRef<HTMLElement>(null);

  const { onMounted } = useMounted();
  const [isOpened, setOpened] = useToggle(open);

  useEffect(() => {
    if (typeof hasNativeDetails === 'undefined') {
      hasNativeDetails =
        typeof document.createElement('details').open === 'boolean';
    }
  }, []);

  useEffect(() => {
    setOpened(open);
  }, [open, setOpened]);

  useEffect(() => {
    if (hasNativeDetails) {
      return;
    }

    if (ref.current) {
      // Don't really need to include this, but just
      // polyfills DOM behaviour for IE/Edge
      if (isOpened) {
        ref.current.setAttribute('open', '');
      } else {
        ref.current.removeAttribute('open');
      }
    }
  }, [isOpened]);

  return (
    <details
      className={classNames('govuk-details', className)}
      open={open}
      ref={ref}
      role={onMounted('group')}
    >
      {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions */}
      <summary
        aria-controls={onMounted(id)}
        aria-expanded={onMounted(isOpened)}
        className="govuk-details__summary"
        role={onMounted('button')}
        tabIndex={onMounted(0)}
        onClick={event => {
          event.persist();

          if (onToggle) {
            onToggle(!isOpened, event);
            if (analytics) {
              analytics({
                category: 'Statistics summary',
                action: `${summary} - ${tag}`,
              });
            }

            if (event.isDefaultPrevented()) {
              return;
            }
          }

          setOpened(!isOpened);
        }}
        onKeyPress={event => {
          if (event.key === 'Enter' || event.key === ' ') {
            event.preventDefault();
            (event.target as HTMLElement).click();
          }
        }}
        onKeyUp={event => {
          if (event.key === ' ') {
            event.preventDefault();
          }
        }}
      >
        <span
          className="govuk-details__summary-text"
          data-testid="details--expand"
        >
          {summary}
          {tag && (
            <span className="govuk-tag govuk-!-margin-left-2">{tag}</span>
          )}
        </span>
      </summary>
      <div
        aria-hidden={onMounted(!isOpened)}
        className="govuk-details__text"
        id={onMounted(id)}
        style={onMounted(
          !hasNativeDetails
            ? {
                display: !isOpened ? 'none' : undefined,
              }
            : undefined,
        )}
      >
        {children}
      </div>
    </details>
  );
};

export default Details;

export const openAllParentDetails = (target: HTMLElement) => {
  const textContainers = findAllParents(target, '.govuk-details__text');

  textContainers.forEach(textContainer => {
    const summary = document.querySelector<HTMLElement>(
      `summary[aria-controls="${textContainer.id}"]`,
    );

    if (summary && summary.getAttribute('aria-expanded') === 'false') {
      summary.click();
    }
  });
};
