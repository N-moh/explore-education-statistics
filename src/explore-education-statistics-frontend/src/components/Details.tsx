import React, { ReactNode } from 'react';

interface Props {
  children: ReactNode;
  summary: string;
}

const Details = ({ children, summary }: Props) => {
  return (
    <details className="govuk-details">
      <summary className="govuk-details__summary">
        <span className="govuk-details__summary-text">{summary}</span>
      </summary>
      <div className="govuk-details__text">{children}</div>
    </details>
  );
};

export default Details;
