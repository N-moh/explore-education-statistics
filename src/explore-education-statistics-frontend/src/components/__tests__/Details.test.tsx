import React from 'react';
import { render } from 'react-testing-library';
import Details from '../Details';

describe('Details', () => {
  test('renders correctly', () => {
    const { container } = render(
      <Details summary="Test summary">
        <p>Test details</p>
      </Details>,
    );

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders correctly with `open` prop set to true', () => {
    const { container } = render(
      <Details summary="Test summary" open>
        <p>Test details</p>
      </Details>,
    );

    expect(container.querySelector('[open]')).not.toBeNull();
    expect(container.innerHTML).toMatchSnapshot();
  });
});
