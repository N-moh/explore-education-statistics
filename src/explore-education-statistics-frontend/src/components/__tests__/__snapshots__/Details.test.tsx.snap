// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details renders correctly 1`] = `

<details class="govuk-details">
  <summary class="govuk-details__summary">
    <span class="govuk-details__summary-text"
          data-testid="details--expand"
    >
      Test summary
    </span>
  </summary>
  <div class="govuk-details__text">
    <p>
      Test details
    </p>
  </div>
</details>

`;

exports[`Details renders correctly with \`open\` prop set to true 1`] = `

<details class="govuk-details"
         open
>
  <summary class="govuk-details__summary">
    <span class="govuk-details__summary-text"
          data-testid="details--expand"
    >
      Test summary
    </span>
  </summary>
  <div class="govuk-details__text">
    <p>
      Test details
    </p>
  </div>
</details>

`;
